File: akka-http/src/main/scala/akka/http/javadsl/server/Unmarshallers.scala
/*
 * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
 */

package akka.http.javadsl.server

import akka.http.impl.server.{ Util, UnmarshallerImpl }
import akka.http.javadsl.model.{ HttpEntity, HttpMessage }
import akka.http.scaladsl.unmarshalling.{ Unmarshaller ⇒ ScalaUnmarshaller, FromMessageUnmarshaller }
import akka.japi.function.Function
import akka.util.ByteString

import scala.reflect.ClassTag

object Unmarshallers {
  def String: Unmarshaller[String] = implicitInstance
  def ByteString: Unmarshaller[ByteString] = implicitInstance
  def ByteArray: Unmarshaller[Array[Byte]] = implicitInstance
  def CharArray: Unmarshaller[Array[Char]] = implicitInstance

  def fromMessage[T](convert: Function[HttpMessage, T], clazz: Class[T]): Unmarshaller[T] =
    new UnmarshallerImpl[T](Util.scalaUnmarshallerFromFunction(convert))(ClassTag(clazz))

  def fromEntity[T](convert: Function[HttpEntity, T], clazz: Class[T]): Unmarshaller[T] =
    new UnmarshallerImpl[T](
      ScalaUnmarshaller.messageUnmarshallerFromEntityUnmarshaller(Util.scalaUnmarshallerFromFunction[HttpEntity, T](convert)))(ClassTag(clazz))

  private def implicitInstance[T: ClassTag](implicit um: FromMessageUnmarshaller[T]): Unmarshaller[T] =
    new UnmarshallerImpl[T](um)
}

File: akka-http/src/main/java/akka/http/javadsl/unmarshalling/Unmarshallers.java
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.http.javadsl.unmarshalling;

import java.util.function.Function;

public interface Unmarshallers {
  /**
   * Creates an unmarshaller from an asynchronous Java function.
   */
  //#unmarshaller-creation
  <A, B> Unmarshaller<A, B> async(java.util.function.Function<A, java.util.concurrent.CompletionStage<B>> f);
  //#unmarshaller-creation

  /**
   * Creates an unmarshaller from a Java function.
   */
  //#unmarshaller-creation
  <A, B> Unmarshaller<A, B> sync(java.util.function.Function<A, B> f);
  //#unmarshaller-creation

  //#unmarshaller-creation
  <A, B> Unmarshaller<A, B> firstOf(Unmarshaller<A, B> u1, Unmarshaller<A, B> u2);
  <A, B> Unmarshaller<A, B> firstOf(Unmarshaller<A, B> u1, Unmarshaller<A, B> u2, Unmarshaller<A, B> u3);
  <A, B> Unmarshaller<A, B> firstOf(Unmarshaller<A, B> u1, Unmarshaller<A, B> u2, Unmarshaller<A, B> u3, Unmarshaller<A, B> u4);
  <A, B> Unmarshaller<A, B> firstOf(Unmarshaller<A, B> u1, Unmarshaller<A, B> u2, Unmarshaller<A, B> u3, Unmarshaller<A, B> u4, Unmarshaller<A, B> u5);
  //#unmarshaller-creation
}
